### ISSUE TITLE
Class cast exception on web-socket cancelation

### ISSUE THREAD

**Issue by altavir** [ktorio/ktor#1929]
**Ktor Version and Engine Used (client or server and name)**
ktor-cio server 1.3.2

**Describe the bug**
I am getting an internal error when force-closing server with opened web-socket connection
Here is the stack trace:
```
13:32:20.839 [DefaultDispatcher-worker-4] ERROR io.ktor.server.cio.HttpServer - Unhandled exception caught for CoroutineName(http-pipeline-writer)
kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for DispatchedContinuation[ExperimentalCoroutineDispatcher@45e6d6a4[scheduler = DefaultDispatcher@6341301e[Pool Size {core = 10, max = 1024}, Worker States {CPU = 6, blocking = 0, parked = 8, dormant = 0, terminated = 0}, running workers queues = [0c, 0c, 0c, 0c, 0c, 0c], global CPU queue size = 5, global blocking queue size = 0, Control State {created workers= 14, blocking tasks = 0, CPUs acquired = 6}]], Continuation at io.ktor.server.cio.backend.ServerPipelineKt$startServerConnectionPipeline$1$outputsActor$1.invokeSuspend(ServerPipeline.kt:49)@2482cf33]. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
	at kotlinx.coroutines.DispatchedTask.handleFatalException$kotlinx_coroutines_core(DispatchedTask.kt:93)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:64)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
Caused by: java.lang.ClassCastException: class kotlin.coroutines.jvm.internal.CompletedContinuation cannot be cast to class kotlinx.coroutines.DispatchedContinuation (kotlin.coroutines.jvm.internal.CompletedContinuation and kotlinx.coroutines.DispatchedContinuation are in unnamed module of loader 'app')
	at kotlinx.coroutines.CoroutineDispatcher.releaseInterceptedContinuation(CoroutineDispatcher.kt:103)
	at kotlin.coroutines.jvm.internal.ContinuationImpl.releaseIntercepted(ContinuationImpl.kt:118)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:39)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:55)
	... 4 common frames omitted
```

**Comment by e5l** [comment]
Hi @altavir, thanks for the report. Could you file a snippet to reproduce the problem?

**Comment by altavir** [comment]
It is not simple to reproduce. Here is the project: https://github.com/mipt-npm/dataforge-control/blob/master/demo/src/main/kotlin/hep/dataforge/control/demo/DemoControllerView.kt. In order to reproduce, you need to show browser window (it opens WebSocket connection automatically), then close the program without closing the browser first.

**Comment by Thomas-Vos** [comment]
Please look into this, @e5l. It looks related to https://github.com/Kotlin/kotlinx.coroutines/issues/1888 and https://github.com/ktorio/ktor/issues/1540. Unfortunately I don't have a simple way to reproduce it. If there is anything I can do to help please let me know.

**Comment by oleg-larshin** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/KTOR-349">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.

**Comment by floscher** [comment]
I also encountered this issue, see https://github.com/Kotlin/kotlinx.coroutines/issues/2413#issuecomment-749022239 and my minimal example: https://github.com/floscher/ktor-ClassCastException

A CI run where the exception occured: https://github.com/floscher/ktor-ClassCastException/runs/1589576823#step:4:10131

**Comment by floscher** [comment]
This seems to be a regression introduced between KTor versions `1.4.1` and `1.4.2`/`1.4.3`:

https://github.com/floscher/ktor-ClassCastException/actions/runs/436231979 vs https://github.com/floscher/ktor-ClassCastException/actions/runs/436232262

**Comment by cy6erGn0m** [comment]
This particular issue is fixed in 1.5.0

**Comment by cy6erGn0m** [comment]
Please file separate tickets for other issues with similar symptoms

**Comment by floscher** [comment]
What I reported above seems to be also fixed with `1.5.0` ( https://github.com/Kotlin/kotlinx.coroutines/issues/2413#issuecomment-751530225 ), so probably had the same underlying cause
